src/kernel.c

```diff
@@ -0,0 +1,640 @@
+	assert(!(th_cur->flags & TH_FLAG_INLIST));
+
+	// Set the current interrupt depth to that of the current thread.
+	__interrupt_depth = th_cur->interrupt_depth;
+
+	return th_cur->stack_state;
+}
+
+/**
+ * @brief Idle thread function
+ *
+ * Since there must always be something run by the CPU, the kernel creates
+ * a "idle" thread: a lowest-priority thread that is run only when all
+ * user-created threads are blocked.
+ *
+ * This thread does absolutely nothing. It just waits for any other thread
+ * to wake up so that it will be preempted.
+ */
+void __kthread_idle(void *arg)
+{
+	// When this thread is scheduled, there are no
+	// ready threads and all threads are probably waiting for some
+	// hardware event to happen. Just spin-wait here.
+	while(1) {}
+}
+
+kthread_t* kernel_init(void)
+{
+	assert(!__kernel);
+	th_ready = NULL;  // empty ready list
+	th_count = 1; // start with the main thread
+
+	// Configure the main thread
+	memset(&th_main, 0, sizeof(th_main));
+	th_main.pri = 0;
+	strlcpy(th_main.name, "main", sizeof(th_main.name));
+
+	// NOTE: keep this in sync with system.c
+	#define STACK_SIZE 0x10000
+	th_main.stack = (char*)0x80000000 + get_memory_size() - STACK_SIZE;
+
+	uint64_t *s = (uint64_t*)th_main.stack;
+	for (int i=0;i<STACK_GUARD/8;i++)
+		s[i] = STACK_COOKIE;
+
+	// The main thread is the currently scheduled one.
+	th_cur = &th_main;
+
+	// Allocate the idle thread
+	th_idle = kthread_new("idle", 4096, -1, __kthread_idle, 0);
+
+	__kernel = true;
+	return th_cur;
+}
+
+void kernel_close(void)
+{
+	assert(__kernel);
+	assertf(th_cur == &th_main, "kthread_close can only be called from main thread");
+
+	// Kill the idle thread to release the memory
+	kthread_kill(th_idle);
+	th_idle = NULL;
+
+	assertf(th_count == 1, "not all threads were killed");
+
+	th_cur = NULL;
+	__kernel = false;
+	__isr_force_schedule = false;
+}
+
+kthread_t* kthread_new(const char *name, int stack_size, int pri, void (*user_entry)(void*), void *user_data)
+{
+	assert((stack_size % 8) == 0);
+
+	// Allocate memory for the stack, the stack guard area, and the kthread_t
+	// structure
+	//
+	// |             |                        |             |
+	// | Stack guard |   Stack                | kthread_t   |
+	// |             |                        |             |
+	//
+	//
+	// For the kthread structure, we will use the final part of
+	// the buffer: since the stack grows downward, this makes sure that
+	// overflowing the stack doesn't corrupt the thread structure, which might
+	// make debugging more difficult and might even cause the kernel to crash.
+	void *thmem = malloc(STACK_GUARD + stack_size + sizeof(kthread_t));
+	kthread_t *th = thmem + STACK_GUARD + stack_size;
+
+	// Initialize the thread structure
+	memset(th, 0, sizeof(kthread_t));
+	th->stack = thmem;
+
+	strlcpy(th->name, name, sizeof(th->name));
+	th->user_entry = user_entry;
+	th->user_data = user_data;
+	th->pri = pri;
```

Comment:
Since negative values aren't allowed, would it be worthwhile to assert if `pri` is less than zero?

Reply:
In theory yes, but I use -1 internally for the idle thread, and I would have to have an internal API to create a thread without this assert. It can be done of course, but I'm not sure it's worth it. It's not that -5 doesn't work, it's just that it's never scheduled because priority is too little. Opinions?

Reply:
Interesting! If that's the case, things are probably fine the way they are. The documentation in `kernel.h` lets the programmer know what's up, and it's not like they'll get a difficult or confusing debugging situation with something like -5. 

Having two initialization functions for threads sounds a little more complicated!